### PR TITLE
Spark date add

### PR DESCRIPTION
**Pull request by RCHowell**
### **date_add**

**Translation**

```
-- partiql
date_add(<part>, <quantity>, <date>)

-- spark
<date> + <interval>
```

**Notes**

The PartiQL date_add is similar to Redshift and Trino’s in which it accepts a datetime part and a quantity. For example, `date_add(months, 2, date)` would add two months to the `date`. Spark only allows adding *days* which means we may need to translate adding years and months; however, this is problematic because of situations like below:

I’ve left out the time part of the timestamp.

* `date_add(month, 1, CAST('2024-01-31' AS TIMESTAMP))`
    * PartiQL → `'2024-02-29'`
    * Trino → `'2024-02-29'`
    * Redshift → `'2024-02-29'`
* `date_add(day, 30, CAST('2024-01-31' AS TIMESTAMP))`
    * PartiQL → `'2024-03-01'`
    * Trino →  `'2024-03-01'`
    * Redshift → `'2024-03-01'`

**Choices**

1. We only allowing adding the `DAYS` datetime part as this is the safest.
2. We make a year = 365 days and a month = 30 days. This can result in errors because a year is 365.25 and a month varies.
3. **CHOSEN: We can do `timestamp  + interval` which allows us to support the PartiQL datetime parts. No obvious drawbacks except perhaps undocumented edge cases?**

**References**

* https://spark.apache.org/docs/2.3.0/api/sql/index.html#date_add
* https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/functions.html#date_add(org.apache.spark.sql.Column,int)
* https://rawgit.com/rstudio/cheatsheets/main/lubridate.pdf

### **date_diff**

**Translation**

```
-- partiql
date_diff(<part>, <d1>, <d2>)

-- spark
<part>(<d2>) - <part>(<d1>)
```

**Notes**

The PartiQL date_diff, given a datetime part and two valid timestamps, returns the difference in datetime parts. 
 On the other hand, Spark’s datediff does not accept a datetime part, and only returns the difference in days.

**Choices**

* **CHOSEN: Extract the datetime part and perform the difference.**
* Only support day and month parts with datediff and months_between

**References**

* https://github.com/partiql/partiql-lang-kotlin/wiki/Functions#date_diff----since-v010
* https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/functions.html#datediff(org.apache.spark.sql.Column,org.apache.spark.sql.Column)
* https://spark.apache.org/docs/2.3.0/api/sql/index.html#datediff

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/test/resources/inputs/builtins/datetime.sql [hunk 1]
```diff
@@ -44,4 +44,22 @@ SELECT DATE_ADD(MONTH, 5, CURRENT_DATE) FROM T;
 SELECT DATE_ADD(YEAR, 5, CURRENT_DATE) FROM T;
 
 --#[datetime-15]
-SELECT DATE_DIFF(DAY, CURRENT_DATE, CURRENT_DATE) FROM T;
+SELECT DATE_DIFF(DAY, timestamp_1, timestamp_2) FROM T;
```
Comment (johnedquinn) [inline src/test/resources/inputs/builtins/datetime.sql:47]: There are now two `datetime-15`

File: src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt [hunk 2]
```diff
@@ -64,4 +69,80 @@ public open class SparkCalls(private val log: ProblemCallback) : SqlCalls() {
         val elementExpr = sqlArgs[2].expr
         return exprCall(fnName, listOf(arrayExpr, elementVar, elementExpr))
     }
+
+    /**
+     * PartiQL: date_add(part: datetime_part, quantity: int, date: date|timestamp) -> date|timestamp`
+     * Spark:   date + interval
+     *
+     * We perform the conversion by a
```
Comment (alancai98) [inline src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt:77]: incomplete comment

**Comment by johnedquinn** [comment]
Reviewed the rest of the PR. Looks good. Just left that one comment.

File: src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt [hunk 3]
```diff
@@ -64,4 +69,80 @@ public open class SparkCalls(private val log: ProblemCallback) : SqlCalls() {
...
+     *  > https://spark.apache.org/docs/2.3.0/api/sql/index.html#date_add
+     *  > https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/functions.html#date_add(org.apache.spark.sql.Column,int)
+     *  > https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/functions.html#add_months(org.apache.spark.sql.Column,int)
+     *  > The number of days to can be negative to subtract days
+     */
+    @OptIn(PartiQLValueExperimental::class)
+    override fun dateAdd(part: DatetimeField, args: SqlArgs): Expr {
+        var quantity = args[0].expr
+        var date = args[1].expr
+        var parts = arrayOfNulls<Expr>(7)
```
Comment (alancai98) [inline src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt:89]: don't think these need to be `var`s

File: src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt [hunk 4]
```diff
@@ -64,4 +69,80 @@ public open class SparkCalls(private val log: ProblemCallback) : SqlCalls() {
...
+
+        // Add detailed warning about this translation.
+        val intervalString = parts.joinToString(
+            prefix = "make_interval(",
+            postfix = ")",
+            separator = ", "
+        ) {
+           if (it != null) "<quantity>" else "0"
+        }
+        log.warn("PartiQL `date_add($part, <quantity>, <date>)` was replaced by Spark `<date> + $intervalString`.")
```
Comment (alancai98) [inline src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt:110]: nit: should this be a `warning` or just an `info`? iirc RedshiftCalls had used an `info` level for its translation of `date_add`.

File: src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt [hunk 5]
```diff
@@ -64,4 +69,80 @@ public open class SparkCalls(private val log: ProblemCallback) : SqlCalls() {
...
+     * PartiQL's date_diff accepts a part whereas Spark's datediff returns the difference in days.
+     *
+     * We extract the datetime part and perform the difference.
+     *
+     * Notes:
+     *  > https://github.com/partiql/partiql-lang-kotlin/wiki/Functions#date_diff----since-v010
+     *  > https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/functions.html#datediff(org.apache.spark.sql.Column,org.apache.spark.sql.Column)
+     *  > https://spark.apache.org/docs/2.3.0/api/sql/index.html#datediff
+     */
+    @OptIn(PartiQLValueExperimental::class)
```
Comment (alancai98) [inline src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt:129]: seems unused?

File: src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt [hunk 6]
```diff
@@ -64,4 +69,80 @@ public open class SparkCalls(private val log: ProblemCallback) : SqlCalls() {
...
+
+    /**
+     * PartiQL's date_diff accepts a part whereas Spark's datediff returns the difference in days.
+     *
+     * We extract the datetime part and perform the difference.
+     *
+     * Notes:
+     *  > https://github.com/partiql/partiql-lang-kotlin/wiki/Functions#date_diff----since-v010
+     *  > https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/functions.html#datediff(org.apache.spark.sql.Column,org.apache.spark.sql.Column)
+     *  > https://spark.apache.org/docs/2.3.0/api/sql/index.html#datediff
```
Comment (alancai98) [inline src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt:127]: Interestingly, `DATE_DIFF` with a space between was added in spark v3.4.0 -- https://spark.apache.org/docs/latest/api/sql/#date_diff.

File: src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt [hunk 7]
```diff
@@ -64,4 +69,80 @@ public open class SparkCalls(private val log: ProblemCallback) : SqlCalls() {
...
+
+        return exprBinary(
+            op = Expr.Binary.Op.PLUS,
+            lhs = date,
+            rhs = interval,
+        )
+    }
+
+    /**
+     * PartiQL's date_diff accepts a part whereas Spark's datediff returns the difference in days.
```
Comment (alancai98) [inline src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt:120]:
```suggestion
     * PartiQL's date_diff accepts a part whereas Spark's datediff returns the difference in days between two dates.
```

File: src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt (hunk 4 above)
Comment (alancai98) [reply src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt:110]: redshiftcalls.kt for reference -- https://github.com/partiql/partiql-scribe/blob/main/src/main/kotlin/org/partiql/scribe/targets/redshift/RedshiftCalls.kt#L42

File: src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt [hunk 8]
```diff
@@ -64,4 +69,80 @@ public open class SparkCalls(private val log: ProblemCallback) : SqlCalls() {
...
+            DatetimeField.MONTH -> "month"
+            DatetimeField.DAY -> "day"
+            DatetimeField.HOUR -> "hour"
+            DatetimeField.MINUTE -> "minute"
+            DatetimeField.SECOND -> "second"
+            else -> error("Unexpected datetime part `$part`")
+        }
+        val d1 = exprCall(id(extract), listOf(args[0].expr))
+        val d2 = exprCall(id(extract), listOf(args[1].expr))
+        return exprBinary(
```
Comment (alancai98) [inline src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt:142]: also could add a log info message about the translation we do for `DATE_DIFF`. RedshiftCalls does something similar -- https://github.com/partiql/partiql-scribe/blob/main/src/main/kotlin/org/partiql/scribe/targets/redshift/RedshiftCalls.kt#L54

**Review by alancai98 — APPROVED** [review]

**Pull request merged**